fem/bilininteg_hdiv.cpp

```diff
@@ -0,0 +1,1987 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license.  We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "../general/forall.hpp"
+#include "bilininteg.hpp"
+#include "gridfunc.hpp"
+#include "libceed/mass.hpp"
+
+using namespace std;
+
+
+// Piola transformation in H(div): w = (1 / det (dF)) dF \hat{w}
+// div w = (1 / det (dF)) \hat{div} \hat{w}
+
+namespace mfem
+{
+
+// PA H(div) Mass Assemble 2D kernel
+void PAHdivSetup2D(const int Q1D,
+                   const int NE,
+                   const Array<double> &w,
+                   const Vector &j,
+                   Vector &_coeff,
+                   Vector &op)
+{
+   const int NQ = Q1D*Q1D;
+   auto W = w.Read();
+
+   auto J = Reshape(j.Read(), NQ, 2, 2, NE);
+   auto coeff = Reshape(_coeff.Read(), NQ, NE);
+   auto y = Reshape(op.Write(), NQ, 3, NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      for (int q = 0; q < NQ; ++q)
+      {
+         const double J11 = J(q,0,0,e);
+         const double J21 = J(q,1,0,e);
+         const double J12 = J(q,0,1,e);
+         const double J22 = J(q,1,1,e);
+         const double c_detJ = W[q] * coeff(q, e) / ((J11*J22)-(J21*J12));
+         // (c/detJ) J^T J
+         y(q,0,e) = c_detJ * (J11*J11 + J21*J21); // 1,1
+         y(q,1,e) = c_detJ * (J11*J12 + J21*J22); // 1,2
+         y(q,2,e) = c_detJ * (J12*J12 + J22*J22); // 2,2
+      }
+   });
+}
+
+// PA H(div) Mass Assemble 3D kernel
+void PAHdivSetup3D(const int Q1D,
+                   const int NE,
+                   const Array<double> &w,
+                   const Vector &j,
+                   Vector &_coeff,
+                   Vector &op)
+{
+   const int NQ = Q1D*Q1D*Q1D;
+   auto W = w.Read();
+   auto J = Reshape(j.Read(), NQ, 3, 3, NE);
+   auto coeff = Reshape(_coeff.Read(), NQ, NE);
+   auto y = Reshape(op.Write(), NQ, 6, NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      for (int q = 0; q < NQ; ++q)
+      {
+         const double J11 = J(q,0,0,e);
+         const double J21 = J(q,1,0,e);
+         const double J31 = J(q,2,0,e);
+         const double J12 = J(q,0,1,e);
+         const double J22 = J(q,1,1,e);
+         const double J32 = J(q,2,1,e);
+         const double J13 = J(q,0,2,e);
+         const double J23 = J(q,1,2,e);
+         const double J33 = J(q,2,2,e);
+         const double detJ = J11 * (J22 * J33 - J32 * J23) -
+         /* */               J21 * (J12 * J33 - J32 * J13) +
+         /* */               J31 * (J12 * J23 - J22 * J13);
```

Comment:
comment leftovers?

Reply:
This was copied and pasted from `VectorMassIntegrator::AssemblePA` which still has these empty comments. It is also in bilininteg_hcurl.cpp, and I will remove it there too for this PR. @jandrej might want to remove them from bilininteg_vecmass.cpp at some point, or I can do it in this PR if desired.

Reply:
I think these are used to trick the `make style`, which would otherwise move the code to the lest.

Reply:
Thanks @vladotomov, I didn't realize that. I just tried `make style` and confirmed that it does move the code to the left, which makes it less readable. I will keep the empty comments.